File: 20/aoc3.cc
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include "aoc_helper.h"

int treesEncountered(std::vector<std::string>data, std::vector<int> path) {
    int horizontal = path[0];
    int down = path[1];
    int tree_count = 0;
    std::vector<int> coordinates = {0,0};
    for(int i = 0; i < data.size(); i++) {
        int current_pos = coordinates[0];
        std::string row = data[i];
        if (row[current_pos] == '#') {
            tree_count++;
        }
        //update coordinates
        coordinates[0] = (current_pos + horizontal) % row.size();
        coordinates[1] += 1;
    }
    return tree_count;
}


int main() {
    //read input from file
    std::vector<std::string> data;
    data = getInputString("./input/aoc3.txt");

    // traverse the map and count the number of trees
    // this path vector means right 3, down 1
    std::vector<int> path = {3, 1};
    int trees = treesEncountered(data, path);
    std::cout << "Encountered " << trees << " trees along the toboggan path" << std::endl;
}

File: 20/python/aoc3.py
# def toboggan(data):
#     trees = 0
#     curr_pos = 0
#     for line in data:
#         line = line.strip()
#         if line[curr_pos] == '#':
#             trees += 1
#         curr_pos = (curr_pos + 3) % len(line)
#     return trees

# if __name__ == "__main__":
#     text_file = open("../input/aoc3.txt", "r")
#     string_data = text_file.readlines()
#     ans = toboggan(string_data)
#     print("the number of trees encountered is {}".format(ans))

def run_toboggan_sim(data, route):
    trees = 0
    curr_pos_x = 0
    curr_pos_y = 0
    while curr_pos_y < len(data):
        line = data[curr_pos_y]
        line = line.strip()
        if line[curr_pos_x] == '#':
            trees += 1
        curr_pos_x = (curr_pos_x + route[0]) % len(line)
        curr_pos_y += route[1]
    return trees

def toboggan2(data):
    trees = 1
    routes = [(1,1), (3,1), (5,1), (7,1), (1, 2)]
    for route in routes:
        trees *= run_toboggan_sim(data, route)
    return trees

if __name__ == "__main__":
    text_file = open("../input/aoc3.txt", "r")
    string_data = text_file.readlines()
    ans = toboggan2(string_data)
    print("the number of trees encountered is {}".format(ans))

   